---
title: "KLRG"
author: "Sung Rye Park"
date: "`r format(Sys.Date())`"
output:  
  rmdformats::robobook: 
    code_folding: hide 
    number_sections: TRUE
    toc_depth: 6
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=F, fig.align = "center", 
                      message=F, warning=F,
                      results = "markup",
                      error = TRUE,
                      highlight = TRUE,
                      prompt = FALSE,
                      tidy = FALSE)
```

# KLRG RNA-seq 
  
Load packages

```{r, echo=TRUE}
library(dplyr)
library(ggplot2)
library(DESeq2)
library(edgeR)
library(reshape)
```

# Import data.   

## Fastq file information   
```{r}
# Create a data frame from the provided data
data_df <- data.frame(
  File = c(
    "20230828_24_hour_tx_DMSO_SJ11376_S54_R1_001.fastq.gz",
    "20230828_24_hour_tx_DMSO_SJ11376_S54_R2_001.fastq.gz",
    "20230828_24_hour_tx_Duvelisib_SJ11376_S53_R1_001.fastq.gz",
    "20230828_24_hour_tx_Duvelisib_SJ11376_S53_R2_001.fastq.gz",
    "20230828_48_hour_tx_DMSO_SJ11376_S56_R1_001.fastq.gz",
    "20230828_48_hour_tx_DMSO_SJ11376_S56_R2_001.fastq.gz",
    "20230828_48_hour_tx_Duvelisib_SJ11376_S55_R1_001.fastq.gz",
    "20230828_48_hour_tx_Duvelisib_SJ11376_S55_R2_001.fastq.gz",
    "20230828_72_hour_tx_DMSO_SJ11376_S58_R1_001.fastq.gz",
    "20230828_72_hour_tx_DMSO_SJ11376_S58_R2_001.fastq.gz",
    "20230828_72_hour_tx_Duvelisib_SJ11376_S57_R1_001.fastq.gz",
    "20230828_72_hour_tx_Duvelisib_SJ11376_S57_R2_001.fastq.gz"
  ),
  MD5 = c(
    "1c3b063bf9d9eda8be1044b1e8fc560d",
    "8606594dbcbd01fa695e71f79a0ed4cb",
    "fdb46ca77f42cb5b2c262b9667a9d0aa",
    "cbb77ecbb652898a3e7b41852e2c735a",
    "80a93d3e6fbf021754f1262a3b5fb65f",
    "e0e94986407d593cde680f7399830fda",
    "0e45e269663dc0c046bc70f08d51c777",
    "9cb15ea7304efeba3bbac10c794cdc82",
    "67eef5df654ed309afd6d166131d3202",
    "22a3b763cc751e77518da86f46d0d0d2",
    "db5cf2bd95b492fe176f815639ed17dd",
    "e07fbd952064f47c0355c2775aca7d86"
  ),
  treatment = c(
    "DMSO", "DMSO", "Duvelisib", "Duvelisib",
    "DMSO", "DMSO", "Duvelisib", "Duvelisib",
    "DMSO", "DMSO", "Duvelisib", "Duvelisib"
  ),
  time = c(
    "24hr", "24hr", "24hr", "24hr",
    "48hr", "48hr", "48hr", "48hr",
    "72hr", "72hr", "72hr", "72hr"
  ),
  R1_R2 = c("R1", "R2", "R1", "R2", "R1", "R2", "R1", "R2", "R1", "R2", "R1", "R2"),
  type = c(
    "RNA-seq", "RNA-seq", "RNA-seq", "RNA-seq",
    "RNA-seq", "RNA-seq", "RNA-seq", "RNA-seq",
    "RNA-seq", "RNA-seq", "RNA-seq", "RNA-seq"
  )
)

# Print the created data frame
print(data_df)
```


Raw counts were generated by STAR method
```{r}
dir='~/Desktop/DF/MGH_Salvia/KLRG/230828_SJ11376_viper5/summary_reports/STAR_Method/'
raw_count = read.csv(paste0(dir,'Counts/STAR_Gene_Counts.csv'), row.names = 1, check.names = F)
raw_count = raw_count[,c(grep("DMSO", colnames(raw_count)), grep("Duvelisib", colnames(raw_count)))]
# raw_count %>% head()
```

## Raw Data

```{r}
cat("Number of features : ", nrow(raw_count), "\n")
cat("Number of samples : ", ncol(raw_count), "\n")
cat("samples : ", paste0(colnames(raw_count), collapse = ','))

```

```{r}
### filtering of features with no expression
cat("features with expression of 0 ","\n")
table(rowSums(raw_count) == 0)
cat("\n")
cat("Number of features : ", nrow(raw_count[rowSums(raw_count) == 0,]), "\n")
raw_count <- raw_count[rowSums(raw_count) != 0,]
cat("Number of features after filtering : ", nrow(raw_count), "\n")
```

```{r}
summary(rowSums(raw_count))
cat("\n")
row_sums <- rowSums(raw_count)
cat("Median raw count of features : ", median(row_sums), "\n")
cat("Mean raw count of features : ", mean(row_sums), "\n")
cat("Minimum raw count of features : ", min(row_sums), "\n")
cat("Maximum raw count of features : ", max(row_sums), "\n")

hist((rowSums(raw_count)), breaks = 1000)
```

```{r, results='markup'}
## input data : info sheet
count.mtx = raw_count
info <- data.frame(matrix(nrow = ncol(count.mtx), ncol = 3))
colnames(info) <- c('sample', 'condition','time')
info$sample <- colnames(count.mtx)
info$condition <- c( rep("DMSO",3),rep("Duvelisib",3))
info$time <- c(rep(c(24,48,72),2))
info$ref = "human_macrophage"
info$concentration= c(rep(0,3), rep("1uM",3))
info
```

```{r}
# define zscore function 
zscore <- function(input.data = input.data){
  input.data.rowsums <- rowSums(input.data)
  input.data.mean <- rowMeans(input.data)
  input.data.sd <- matrixStats::rowSds(as.matrix(input.data))
  names(input.data.sd) <- rownames(input.data)
  zscore <- (input.data-input.data.mean)/input.data.sd
  return(zscore)
}
```


# Raw count 

We can do pairwise comparisons at those timepoints and even if there was one timepoint at which we can demonstrate repolarization of macrophages with duvelisib to M1 phenotype, it would be bingo as we have corroborating ADCP, cytokine, apoptosis and flow/phenotypic data. 
   
Q. what is ADCP?   
   
   
Genes to investigate    
```{r}
gene_vector <- c(
  "CD86", "CIITA", "HLA-DMA", "HLA-DMB", "HLA-DPB1", "HLA-DRB5",
  "CXCL2", "CXCL3", "CXCL8", "TLR4", "TLR5", "TLR7", "TLR8",
  "LRP1", "CD14", "CD44", "FCGR2A", "FCGR3A", "GAS6", "MFGE8",
  "CD300A", "SIRPA", "SIGLEC10", 
  "LYZ", "CTSS", 
  "IL10","CD274", "CD276"
)
```

```{r}
genes = gene_vector[gene_vector %in% rownames(raw_count)]
genes 
```

```{r}
scaled_raw_count <- zscore(input.data  = raw_count[genes,])
```


## scaled raw count heatmap with selected genes 
```{r, fig.height=8, fig.width=4}
#genes = gene_vector[gene_vector %in% rownames(scaled_raw_count)]
scaled_raw_count[genes,] %>% pheatmap::pheatmap(cluster_cols = F, cluster_rows = F)
```

# TPM
```{r}
dir='~/Desktop/DF/MGH_Salvia/KLRG/230828_SJ11376_viper5/summary_reports/STAR_Method/'
tpm = read.csv(paste0(dir,'Counts/STAR_Gene_TPM.csv'), row.names = 1, check.names = F)
tpm = tpm[,c(grep("DMSO", colnames(tpm)), grep("Duvelisib", colnames(tpm)))]
tpm = tpm[,c(grep("DMSO", colnames(tpm)), grep("Duvelisib", colnames(tpm)))]
tpm %>% head()
```


```{r}
cat("Number of features : ", nrow(tpm), "\n")
cat("Number of samples : ", ncol(tpm), "\n")
cat("samples : ", paste0(colnames(tpm), collapse = ','))

```

```{r}
### filtering of features with no expression
table(rowSums(tpm) == 0)
cat("\n")
cat("Number of features : ", nrow(tpm[rowSums(tpm) == 0,]), "\n")
tpm <- tpm[rowSums(tpm) != 0,]
cat("Number of features : ", nrow(tpm), "\n")
```


Genes to investigate.  
```{r}
gene_vector <- c(
  "CD86", "CIITA", "HLA-DMA", "HLA-DMB", "HLA-DPB1", "HLA-DRB5",
  "CXCL2", "CXCL3", "CXCL8", "TLR4", "TLR5", "TLR7", "TLR8",
  "LRP1", "CD14", "CD44", "FCGR2A", "FCGR3A", "GAS6", "MFGE8",
  "CD300A", "SIRPA", "SIGLEC10", 
  "LYZ", "CTSS", 
  "IL10","CD274", "CD276"
)
```

```{r}
genes = gene_vector[gene_vector %in% rownames(tpm)]
genes 
```

```{r}
scaled_tpm <- zscore(input.data  = tpm[genes,])
```

## scaled tpm heatmap with selected genes 
```{r, fig.height=8, fig.width=4}
scaled_tpm[genes,] %>% pheatmap::pheatmap(cluster_cols = F, cluster_rows = F)
```


# FPKM

## FPKM calculation
```{r}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(GenomicFeatures)
geneset <- rownames(raw_count) ## gene list

# Getting genomic information 
allGenes <- genes(TxDb.Hsapiens.UCSC.hg19.knownGene)
allGenes.df <- allGenes %>% data.frame()

allGenes_ranges <- data.frame(allGenes@ranges)
allGenes_ranges$ENTREZID <- allGenes_ranges$names


# Convert ENTREZID to symbol
genes_to_convert <- clusterProfiler::bitr(geneset, fromType = "SYMBOL", 
                                          toType = c("ENSEMBL","ENTREZID"), 
                                          OrgDb = "org.Hs.eg.db")

## remove duplicated genes (Symbol)
genes_to_convert <- genes_to_convert[!duplicated(genes_to_convert$SYMBOL),]
rownames(genes_to_convert) <- genes_to_convert$SYMBOL

## subset out the selected genes (genes from RNA-seq data) range information

selected_genes <- allGenes_ranges[allGenes_ranges$names %in% genes_to_convert$ENTREZID,]

## two inputs
## genes to convert : symbol, ENTREZID
## selected genes : ENTREZID, location, width 
## join columns by ENTREZID
gene_info <- left_join(genes_to_convert, selected_genes, by='ENTREZID')

# merge information
df.merge <- merge(gene_info[,c('SYMBOL','names')], 
                  allGenes.df[,c('seqnames','start','end','width','gene_id')], by.x='names',by.y='gene_id')
# table(is.na(df.merge)) : no NA
# df.merge[1:3,]
# table(duplicated(df.merge$SYMBOL))
# table(duplicated(df.merge$names))

df.merge <- df.merge[!(duplicated(df.merge$SYMBOL)),]

rownames(df.merge) <- df.merge$SYMBOL
genes <- rownames(raw_count)[rownames(raw_count) %in% df.merge$SYMBOL]
raw_count <- raw_count[genes,]
# generate GRanges object
raw_count.gr <- df.merge[genes,][,c(3:6,1:2)] %>% makeGRangesFromDataFrame()

# generate dds object
se <- SummarizedExperiment(as.matrix(raw_count), 
                           colData=DataFrame(sample=1:ncol(count.mtx)))
dds <- DESeqDataSet(se, ~ 1)
rowRanges(dds) <- raw_count.gr

# the FPKM values
fpkm <- fpkm(dds) ## genomic info가 있어야함
# fpkm[1:3,]
# fpkm %>% nrow()
```



```{r}
scaled_fpkm <- scale(fpkm, center = TRUE, scale = TRUE)
min_value <- -1
max_value <- 1
scaled_fpkm <- (scaled_fpkm - min(scaled_fpkm)) / (max(scaled_fpkm) - min(scaled_fpkm)) * (max_value - min_value) + min_value
```

Genes to investigate.  
```{r}
gene_vector <- c(
  "CD86", "CIITA", "HLA-DMA", "HLA-DMB", "HLA-DPB1", "HLA-DRB5",
  "CXCL2", "CXCL3", "CXCL8", "TLR4", "TLR5", "TLR7", "TLR8",
  "LRP1", "CD14", "CD44", "FCGR2A", "FCGR3A", "GAS6", "MFGE8",
  "CD300A", "SIRPA", "SIGLEC10", 
  "LYZ", "CTSS", 
  "IL10","CD274", "CD276"
)
```

```{r}
genes = gene_vector[gene_vector %in% rownames(fpkm)]
genes 
```

```{r}
scaled_fpkm <- zscore(input.data  = fpkm[genes,])
```

## scaled fpkm heatmap with selected genes 
```{r, fig.height=8, fig.width=4}
scaled_fpkm[genes,] %>% pheatmap::pheatmap(cluster_cols = F, cluster_rows = F)
```


# DEGs 

control: DMSO 24, 48, 72 merged  
treat: duvelisib 24, 48, 72 merged   
```{r}
dds <- DESeqDataSetFromMatrix(raw_count, info, ~ condition)
cat("DESeq object dimension" ,"\n")
dim(dds)
dds <- DESeq(dds)
res <- results(dds)
cat("DEGs object dimension" ,"\n")
dim(res)
res <- data.frame(res)
```

```{r}
## volcano plot
cat("p value is not significant in this comparison", "\n")
res %>% ggplot(aes(log2FoldChange, -log10(pvalue))) + geom_point(alpha=0.5) +
  geom_vline(xintercept = c(-log2(1.2), log2(1.2)), color='red')
```

## top 50 upregulated in treatment (by log2FoldChange).  
```{r}
top50_positive =res %>% top_n(50, log2FoldChange) %>% rownames()
cat("top 50 upregulated in treatment : ", paste0(top50_positive, collapse = ","), "\n")
```


## top 50 downregulated in treatment (by log2FoldChange).  
```{r}
top50_negative =res %>% top_n(-50, log2FoldChange) %>% rownames()
cat("top 50 downregulated in treatment : ", paste0(top50_negative, collapse = ","), "\n")
```

## top 50 upregulated in treatment information   
```{r}
merged_df <- merge(res[top50_positive,], raw_count[top50_positive,], by = "row.names", all = TRUE) 
colnames(merged_df)[1] <- "Row.names"
rownames(merged_df) = merged_df$Row.names
merged_df = merge(merged_df, tpm[top50_positive,], by = "row.names", all = TRUE) 
colnames(merged_df)[1:2] = c("gene in raw count", "gene in tpm")
colnames(merged_df) <- sub("\\.x$", ".raw_count", colnames(merged_df))
colnames(merged_df) <- sub("\\.y$", ".tpm", colnames(merged_df))
merged_df = merged_df[merged_df$`gene in raw count` %in% top50_positive,]
```

```{r}
merged_df %>% DT::datatable(extensions = 'Buttons',
            options = list(dom = 'Blfrtip',
                           buttons = c('csv', 'excel')))
```


## top 50 downregulated in treatment information   
```{r}
merged_df <- merge(res[top50_negative,], raw_count[top50_negative,], by = "row.names", all = TRUE) 
colnames(merged_df)[1] <- "Row.names"
rownames(merged_df) = merged_df$Row.names
merged_df = merge(merged_df, tpm[top50_negative,], by = "row.names", all = TRUE) 
colnames(merged_df)[1:2] = c("gene in raw count", "gene in tpm")
colnames(merged_df) <- sub("\\.x$", ".raw_count", colnames(merged_df))
colnames(merged_df) <- sub("\\.y$", ".tpm", colnames(merged_df))
merged_df = merged_df[merged_df$`gene in raw count` %in% top50_negative,]
```

```{r}
merged_df %>% DT::datatable(extensions = 'Buttons',
            options = list(dom = 'Blfrtip',
                           buttons = c('csv', 'excel')))
```
